From 12ee10bee868b1db08ebc8127af8f666ca46f328 Mon Sep 17 00:00:00 2001
From: nikhilad <nikhilad@ami.com>
Date: Wed, 29 Oct 2025 18:14:41 +0530
Subject: [PATCH] Added Test-Fru Feature

[Recreated the Patch]

Support for Test FRU (TestFru dbus object)

Async systemd watchdog notify (120s)

Support for TMP432, TMP461, MAX16550, TMP1075 drivers

EEPROM detection method modified to avoid FPGA crash.

Tested. working as expected.

Signed-off-by: saravananm <saravananm@ami.com>

Signed-off-by: nikhilad <nikhilad@ami.com>
---
 meson.options                  |  7 +++++++
 src/entity_manager/devices.hpp | 17 +++++++++++++++++
 src/fru_device/fru_device.cpp  | 32 ++++++++++++++++++++++++++++++++
 src/meson.build                |  4 ++++
 4 files changed, 60 insertions(+)

diff --git a/meson.options b/meson.options
index 65c1f97..7df1d36 100644
--- a/meson.options
+++ b/meson.options
@@ -28,3 +28,10 @@ option(
     value: true,
     description: 'Run JSON schema validation during the build.',
 )
+option(
+    'test-fru',
+    value : false,
+    type: 'boolean',
+    description: 'On target Mock FRU device for test',
+)
+
diff --git a/src/entity_manager/devices.hpp b/src/entity_manager/devices.hpp
index 28028bb..0dabefb 100644
--- a/src/entity_manager/devices.hpp
+++ b/src/entity_manager/devices.hpp
@@ -199,6 +199,23 @@ const boost::container::flat_map<const char*, ExportTemplate, CmpStr>
           ExportTemplate("raa229621 $Address", "/sys/bus/i2c/devices/i2c-$Bus",
                          "new_device", "delete_device",
                          createsHWMon::hasHWMonDir)},
+        {"TMP432",
+
+          ExportTemplate("tmp432 $Address", "/sys/bus/i2c/devices/i2c-$Bus",
+                         "new_device", "delete_device",
+                         createsHWMon::hasHWMonDir)},
+         {"TMP461",
+          ExportTemplate("tmp461 $Address", "/sys/bus/i2c/devices/i2c-$Bus",
+                        "new_device", "delete_device",
+                         createsHWMon::hasHWMonDir)},
+         {"TMP1075",
+          ExportTemplate("tmp1075 $Address", "/sys/bus/i2c/devices/i2c-$Bus",
+                         "new_device", "delete_device",
+                         createsHWMon::hasHWMonDir)},
+         {"MAX16550",
+          ExportTemplate("max16550 $Address", "/sys/bus/i2c/devices/i2c-$Bus",
+                          "new_device", "delete_device",
+                          createsHWMon::hasHWMonDir)},
          {"PIC32",
           ExportTemplate("pic32 $Address", "/sys/bus/i2c/devices/i2c-$Bus",
                          "new_device", "delete_device",
diff --git a/src/fru_device/fru_device.cpp b/src/fru_device/fru_device.cpp
index 960ca68..a30d776 100644
--- a/src/fru_device/fru_device.cpp
+++ b/src/fru_device/fru_device.cpp
@@ -562,6 +562,30 @@ int getBusFRUs(int file, int first, int last, int bus,
 
         // Scan for i2c eeproms loaded on this bus.
         std::set<size_t> skipList = findI2CEeproms(bus, devices);
+        if (!skipList.empty())
+        {
+            // NOTE:
+            // In this function there are two ways to find the EEPROM.
+            // and record the information in the devices data structure.
+            // Method 1. In findI2CEeproms function,
+            //    it search "/sys/bus/i2c/devices/" and find the EEPROM device.
+            // Method 2. Find out the slave address of all devices
+            //    on the specified I2C bus, and test whether device
+            //    is 8 or 16 bits EEPROM by sending data 0 to each device.
+            //    Unfortunately,
+            //    FPGA Soc will crash when it receives 0 data,
+            //    and it will not return to normal until the power cycle.
+            // According to the experimental results,
+            // the Method 1 can find out all the EEPROMs on the system
+            // on a specific I2C bus and record the information,
+            // in this case,
+            // There is no need to proceed with Method 2.
+
+            // Find the EEPROM on the specified I2C bus,
+            // and record the information in the devices data structure
+            return 1;
+        }
+
         std::set<size_t>& failedItems = failedAddresses[bus];
         std::set<size_t>& foundItems = fruAddresses[bus];
         foundItems.clear();
@@ -1662,6 +1686,14 @@ int main()
     */
     iface->initialize();
 
+#ifdef ENABLE_TEST_FRU
+    std::shared_ptr<sdbusplus::asio::dbus_interface> testIface =
+        objServer.add_interface("/xyz/openbmc_project/FruDevice/TestFru",
+                                "xyz.openbmc_project.FruDevice");
+    testIface->register_property("BOARD_PRODUCT_NAME", 0);
+    testIface->initialize();
+#endif // ENABLE_TEST_FRU
+
     std::function<void(sdbusplus::message_t & message)> eventHandler =
         [&](sdbusplus::message_t& message) {
             std::string objectName;
diff --git a/src/meson.build b/src/meson.build
index 2f98faf..7120835 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -8,6 +8,10 @@ if get_option('fru-device')
     subdir('fru_device')
 endif
 
+if get_option('test-fru')
+    cpp_args_fd = cpp_args_fd + ['-DENABLE_TEST_FRU']
+endif
+
 if get_option('devicetree-vpd')
     subdir('devicetree_vpd_parser')
 endif
-- 
2.34.1

